### PR TITLE
[fix](window_funnel) fix wrong result of window_funnel

### PR DESCRIPTION
**Pull request by jacktengg**
## Proposed changes

Issue Number: close #xxx

Current logic of `window_funnel` is wrong, it cannot express the semantic of the function.

This PR re-implement the logic.

### CODE REVIEW

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run performance

**Comment by jacktengg** [comment]
run cloud_p0

**Comment by jacktengg** [comment]
run performance

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h [hunk 1]
```diff
@@ -72,138 +80,233 @@ WindowFunnelMode string_to_window_funnel_mode(const String& string) {
...
+            column->clear();
+        }
+        block.clear_column_data();
+        sorted = false;
+        is_merge = false;
     }
 
-    void add(const DateValueType& timestamp, int event_idx, int event_num, int64_t win,
-             WindowFunnelMode mode) {
+    void add(const IColumn** arg_columns, ssize_t row_num, int64_t win, WindowFunnelMode mode) {
```
Comment (mrhhsg) [inline be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:129]: why do we need to use `ssize_t` here?

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h [hunk 2]
```diff
@@ -72,138 +80,233 @@ WindowFunnelMode string_to_window_funnel_mode(const String& string) {
...
 
-        if (sorted && events.size() > 0) {
-            if (events.back().first == timestamp) {
-                sorted = events.back().second <= event_idx;
-            } else {
-                sorted = events.back().first < timestamp;
-            }
+        timestamp_column->insert_from(*arg_columns[2], row_num);
+        for (int i = 0; i < event_count; i++) {
+            event_columns[i]->insert_from(*(arg_columns[3 + i]), row_num);
```
Comment (mrhhsg) [inline be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:135]: Calling `insert_from`(virtual function) row by row may lead to performance issues.

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h [hunk 3]
```diff
@@ -72,138 +80,233 @@ WindowFunnelMode string_to_window_funnel_mode(const String& string) {
...
-                            events_timestamp[event_idx].value().second > timestamp) {
-                            events_timestamp[event_idx] = {first_timestamp, timestamp};
-                        }
-                        if (event_idx + 1 == max_event_level) {
-                            // Usually, max event level is small.
-                            return max_event_level;
+                    break;
+                }
+                const auto& event_column = block.get_by_position(column_idx);
+                const auto& event_data =
```
Comment (mrhhsg) [inline be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:205]: duplicated with line 190,191

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h [hunk 4]
```diff
@@ -72,138 +80,233 @@ WindowFunnelMode string_to_window_funnel_mode(const String& string) {
...
+                        assert_cast<const ColumnVector<UInt8>&>(*event_column.column).get_data();
+                match_row = simd::find_one(event_data.data(), row_count, match_row + 1);
+                if (match_row < row_count) {
+                    auto current_timestamp =
+                            binary_cast<NativeType, DateValueType>(timestamp_data[match_row]);
+                    bool is_matched = current_timestamp <= end_timestamp;
+                    if (is_matched) {
+                        if constexpr (WINDOW_FUNNEL_MODE == WindowFunnelMode::INCREASE) {
```
Comment (mrhhsg) [inline be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:213]: Maybe we should merge this with lines 220-223

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h [hunk 5]
```diff
@@ -72,138 +80,233 @@ WindowFunnelMode string_to_window_funnel_mode(const String& string) {
...
+                                 tmp_column_idx++) {
+                                const auto& tmp_event_column =
+                                        block.get_by_position(tmp_column_idx);
+                                const auto& tmp_event_data =
+                                        assert_cast<const ColumnVector<UInt8>&>(
+                                                *tmp_event_column.column)
+                                                .get_data();
+                                auto dup_match_row = simd::find_one(
+                                        tmp_event_data.data(), rows_to_match, last_match_row + 1);
+                                if (dup_match_row >= rows_to_match) {
```
Comment (mrhhsg) [inline be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:238]: `dup_match_row >= rows_to_match` means not matched?

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h [hunk 6]
```diff
@@ -212,64 +315,97 @@ struct WindowFunnelState {
...
                           out);
         }
-        write_var_int(events.size(), out);
-
-        for (int64_t i = 0; i < events.size(); i++) {
-            int64_t timestamp = binary_cast<DateValueType, NativeType>(events[i].first);
-            int event_idx = events[i].second;
-            write_var_int(timestamp, out);
-            write_var_int(event_idx, out);
+        PBlock pblock;
```
Comment (mrhhsg) [inline be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:327]: Will there be compatibility issues here?

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h (hunk 1 above)
Comment (jacktengg) [reply be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:129]: ```row_num```  of ```AggregateFunctionWindowFunnel::add``` is ```ssize_t```

**Comment by jacktengg** [comment]
run buildall

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h (hunk 6 above)
Comment (jacktengg) [reply be/src/vec/aggregate_functions/aggregate_function_window_funnel.h:327]: It's inevitable, the previous data structure cannot express the semantic of window_funnel.

**Comment by jacktengg** [comment]
run buildall

File: be/src/vec/aggregate_functions/aggregate_function_window_funnel.h (hunk 5 above)
Comment (jacktengg) [reply be/src/vec/aggregate_functions/aggregate_function_window_funnel.h]: fixed

**Review by mrhhsg — DISMISSED** [review]
LGTM

**Comment by jacktengg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by mrhhsg — APPROVED** [review]
lgtm

**Pull request merged**